modules/cache/logic/cacheLogic.js

```diff
@@ -107,6 +108,83 @@ module.exports = class CacheLogic {
+      async () => aeternity.fungibleTokenTotalSupply(wordSaleState.token), cache.shortCacheTime);
+
+    const price = cache.getOrSet(['wordSalePrice', address],
+      () => aeternity.wordSalePrice(address), cache.shortCacheTime);
+
+    const [buy, sell] = await price;
+    return {
+      wordSaleAddress: address,
+      tokenAddress: wordSaleState.token,
+      totalSupply: await totalSupply,
+      buyPrice: buy,
+      sellPrice: sell,
+      spread: wordSaleState.spread,
+      description: wordSaleState.description,
+    };
+  }
+
+  static async wordSaleDetailsByToken(address) {
+    const wordRegistryData = await CacheLogic.getWordRegistryData();
+    const wordDetails = await wordRegistryData.tokens.asyncMap(([, wordSale]) => CacheLogic.wordSaleDetails(wordSale));
+    return wordDetails.find(sale => sale.tokenAddress === address);
+  }
+
+  static async wordSaleVotesDetails(address) {
+    const votes = await cache.getOrSet(['wordSaleVotes', address],
+      () => aeternity.wordSaleVotes(address), cache.shortCacheTime);
+
+    return Promise.all(votes.map(([id, vote]) => CacheLogic.wordSaleVoteInfo(id, vote[1], vote[0], address)));
+  }
+
+  static async wordSaleVoteInfo(id, vote, alreadyApplied, sale) {
+    const state = cache.getOrSet(['wordSaleVoteState', vote],
+      () => aeternity.wordSaleVoteState(vote), cache.shortCacheTime);
+
+    const voteTimeout = cache.getOrSet(['wordSaleVoteTimeout', sale],
+      () => aeternity.wordSaleVoteTimeout(sale));
+
+    const height = aeternity.client.height();
+
+    const votedFor = (await state).vote_state.find(([s]) => s)[1];
+    const votedAgainst = (await state).vote_state.find(([s]) => !s)[1];
+    const ifAgainstZero = votedFor === 0 ? 0 : 100;
+    const votedPositive = new BigNumber(votedFor)
+      .dividedBy(new BigNumber(votedFor).plus(votedAgainst)).times(100).toFixed(0);
+
+    const tokenAddress = cache.getOrSet(['wordSaleTokenAddress', sale],
+      () => aeternity.wordSaleTokenAddress(sale));
+
+    const totalSupply = cache.getOrSet(['fungibleTokenTotalSupply', await tokenAddress],
+      async () => aeternity.fungibleTokenTotalSupply(await tokenAddress), cache.shortCacheTime);
+
+    const stakePercent = (await totalSupply) ? new BigNumber(votedFor).dividedBy(await totalSupply).times(100).toFixed(0) : '0';
+    const timeoutHeight = (await state).close_height + (await voteTimeout);
+    return {
+      id,
+      alreadyApplied,
+      voteAddress: vote,
+      subject: (await state).metadata.subject,
+      timeouted: timeoutHeight < (await height),
+      timeoutHeight,
+      closeHeight: (await state).close_height,
+      voteAccounts: (await state).vote_accounts,
+      isClosed: (await height) >= (await state).close_height,
+      isSuccess: new BigNumber(stakePercent).isGreaterThan(50),
+      votePercent: votedAgainst !== 0 ? votedPositive : ifAgainstZero,
+      stakePercent,
+    };
+  }
```

Comment:
should we move this to the token logic? seems like a lot more than caching here

Reply:
the problem ist the call is needed in CacheLogic currently, to trigger fetch again after invalidation, we can move the logic to token logic when we use the message queue here I think

Reply:
or move invalidation routes to tokenLogic

Reply:
Okay then lets add a todo here and then this is to be done when we merge message queues